### PR TITLE
Refactor/ptl model

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This is part 3 of a `chemprop` refactor that ties together both parts 1 (#282) and 2 (#326) into a full system that allows for training the models.

## Current paradigm
Models are currently trained/used for prediction via the [`chemprop.train`](https://github.com/chemprop/chemprop/tree/master/chemprop/train) module. This code can be described broadly as both (i) boilerplate and (ii) convoluted. Further description is below

### 1. "boilerplate"
at the end of the day, `chemprop` is about training a PyTorch model, which really all follows the general structure:
```python
for _ in range(epochs):
    model.train()
    for batch in train_loader:
        model.zero_grad()
        inputs, targets = batch
        preds = model(inputs)
        loss = criterion(preds, targets)
        loss.backward()
        optimizer.step()

    model.eval()
    for batch in val_loader():
        inputs, targets = batch
        preds = model(inputs)
        loss = metric(preds, targets)
        # report metric
```
A lot of this general structure exists in chemprop, but it doesn't need to. [PyTorch Lightning](https://pytorch-lightning.readthedocs.io/en/stable/) removes the need for most of the boilerplate involved in PyTorch model training, cutting away a lot of repeated code.

### 2. "convoluted"
If only model training in `chemprop` were truly as simple as I depicted above. The reality is that at train time, we need to know both the *type* of data on which were training (regression, classification, multiclass, spectral) and *how* it is being trained (i.e., the loss function inputs like bounded MSE, Dirichlet, etc.) Currently, `chemprop` model training is "agnostic" as to these distinctions, and by that I mean that they're all wrapped inside one function, [`train`](https://github.com/chemprop/chemprop/blob/master/chemprop/train/train.py#L17), which then determines how to treat the data. Realistically, this offers no benefits and only serves to (1) make the code impenetrable to readers/maintainers and (2) slower to run (as it must constantly flow through a bunch of branched logic to execute correctly.)

As an example, can anyone easily tell me what the block below is trying to do?
```
       if args.loss_function == 'mcc' and args.dataset_type == 'classification':
            loss = loss_func(preds, targets, data_weights, mask) *target_weights.squeeze(0)
        elif args.loss_function == 'mcc': # multiclass dataset type
            targets = targets.long()
            target_losses = []
            for target_index in range(preds.size(1)):
                target_loss = loss_func(preds[:, target_index, :], targets[:, target_index], data_weights, mask[:, target_index]).unsqueeze(0)
                target_losses.append(target_loss)
            loss = torch.cat(target_losses).to(torch_device) * target_weights.squeeze(0)
        elif args.dataset_type == 'multiclass':
            targets = targets.long()
            if args.loss_function == 'dirichlet':
                loss = loss_func(preds, targets, args.evidential_regularization) * target_weights * data_weights * mask
            else:
                target_losses = []
                for target_index in range(preds.size(1)):
                    target_loss = loss_func(preds[:, target_index, :], targets[:, target_index]).unsqueeze(1)
                    target_losses.append(target_loss)
                loss = torch.cat(target_losses, dim=1).to(torch_device) * target_weights * data_weights * mask
        elif args.dataset_type == 'spectra':
            loss = loss_func(preds, targets, mask) * target_weights * data_weights * mask
        elif args.loss_function == 'bounded_mse':
            loss = loss_func(preds, targets, lt_target_batch, gt_target_batch) * target_weights * data_weights * mask
        elif args.loss_function == 'evidential':
            loss = loss_func(preds, targets, args.evidential_regularization) * target_weights * data_weights * mask
        elif args.loss_function == 'dirichlet': # classification
            loss = loss_func(preds, targets, args.evidential_regularization) * target_weights * data_weights * mask
        else:
            loss = loss_func(preds, targets) * target_weights * data_weights * mask
```

## Updated code
This PR aims to address both of these current limitations through the use of PyTorch lightning and by separating different models into classes. Abstractly, chemprop defines the following "types" of models:

- regression models:
- classification models
- multiclass models
- spectral prediction models

and the current PR has defined each of these as an extension of a base multi-task `MPNN` model that knows how it (i) trained and (ii) used for inference.  To the extent possible, each model is expressed as an extension of the base multi-task `MPNN`, with the main differences in each subclass being the prediction logic. By-and-large, the models are all trained the "same" way, with the differences being handled inside a new `LossFunction` object, which takes the raw predictions and does whatever transformations necessary before calculating a loss.

For example, consider the following single-task models:
1.  regression
2. binary classification model

Abstractly, both of these produce a single scalar prediction for each with the raw predictions of (1) corresponding to regression target values and (2) the logits. During training, the loss is calculated with respect to the target and these raw predictions then backpropagated. The thing that actually separates a regression/classification model is not _how_ they're trained, but specifically the _loss_ that is used to train them (e.g., MSE vs BCE). They are also separated by how they are used for inference, with classification model outputs being subject to a `sigmoid` transformation.

## Questions
I'm still not sold on this precise organization, but I can't think of a better way. There's the broad question though of "allowed" loss functions. Currently, chemprop forbids mixing loss functions between dataset types. That is, you can't declare a model to be a regression model then hand it a BCE loss. This is maybe nice from a client perspective and preventing accidents, but I can't argue why this is fundamentally wrong. The only cases where loss mismatches should be raising errors would be for multi-{target,class} losses (e.g., MVE, Evidential, cross entropy), where the loss calculation would inevitably raise an error. **However**, the reason this has not been done is that I want to offer the flexibility of a user inputting their own custom `LossFunction`. If we perform validation of the loss function at model initialization, that requires then that the user modify the source of the `MPNN` validation code to accept their new loss function, something that I'd like to avoid as it's pretty hacky. Maybe the solution right now is just to raise a warning?

## Checklist
- [x] linted with flake8?
- [ ] (if appropriate) unit tests added? **no but this requires a larger conversation about how to effectively unit test models**

### CODE REVIEW

**Comment by davidegraff** [comment]
this PR isn't ready yet- still some issues with MCC
